### PR TITLE
[BREAKING][ckpt, trainer] refactor: extract ModelCheckpointManager and slim checkpoint callbacks

### PR DESCRIPTION
**Pull request by Coach257**
### What does this PR do?

Split checkpoint I/O from cadence, without introducing `VeOmniModelRuntime`.

| Layer | Owner |
|---|---|
| **When** | `CheckpointCallback` (DCP + HF/LoRA cadence, plus one-shot tokenizer/config sidecars) |
| **What** | `BaseTrainer.load` / `save_dcp` / `save_hf_or_lora` / `save_model_assets` |
| **How** | `ModelCheckpointManager(trainer)` — DCP / HF / LoRA, drain-async, `empty_cache`, barrier, directory layout |
| **Job cursor** | `GlobalStateCallback` — per-rank `trainer_state_rank_{N}.pt` (dataloader, rng, meters, channel-loss) |

DCP `extra_state` is **only** `lr_scheduler`. Staging still keys on the **run-root** `save_path` plus `global_steps=state.global_step` (folding the step into the path strands per-step staging dirs).

**Removed public names:** `CheckpointerCallback`, `HuggingfaceCkptCallback`, `HFLoraCkptCallback`. One `CheckpointCallback`; the manager branches on `lora_config`.

Independent of the mesh-from-config split. After both land, merge `main` into #1125 so that PR is runtime-only.

### Checklist Before Starting

- Search for relative PRs/issues and link here: split out of #1125. Companion: `szl.init-parallel-state-from-config`. Coach257 on #1125: fold extra checkpoint CI into `test_checkpoint_callback.py`.
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

Extended the already CI-enumerated `tests/checkpoints/test_checkpoint_callback.py` (cadence, sidecar-then-load on train begin, staging path, extra_state is scheduler-only, GlobalStateCallback channel-loss, legacy extra_state restore). GPU/NPU unit jobs run that file and `test_dcp_checkpointer.py` in one pytest invocation. `test_trainer_saveload.py` stays a separate e2e step (CUDA-context isolation). Did **not** add `test_model_checkpoint_manager.py` / `test_global_state_callback.py` workflow lines.

Also retargeted `test_trainer_saveload.py`, data resume helpers, and LoRA trainer tests to the new callback names. Production `BaseTrainer._init_callbacks` installs `GlobalStateCallback`; tests that override `_init_callbacks` and still need step/dataloader resume do too.

CPU: `pytest tests/checkpoints/test_checkpoint_callback.py` — 19 passed. `make quality` clean.

### API and Usage Example

**Before:** `CheckpointerCallback` both scheduled and wrote DCP, stuffing job state into `extra_state`. `HuggingfaceCkptCallback` / `HFLoraCkptCallback` were two classes.

**After:**

```python
# BaseTrainer._init_callbacks
self.checkpoint = ModelCheckpointManager(self)
self.checkpoint_callback = CheckpointCallback(self)  # assets / load / save_dcp / save_hf_or_lora
self.global_state_callback = GlobalStateCallback(self)

# DCP save contract (run-root + global_steps, not a per-step path)
checkpointer.save(self.config.save_path, state, global_steps=state.global_step, stage_dir=self.config.stage_dir)
```

On-disk (single-model job):

```
<save_path>/global_step_{N}/
  ├── __0_0.distcp …          # model + optimizer + {lr_scheduler}
  ├── trainer_state_rank_{R}.pt
  └── hf_ckpt/                # optional HF export
```

Checkpoints written by the old `CheckpointerCallback` still restore the job cursor from `extra_state` (compat path in `ModelCheckpointManager._load_extra_state`). New saves do not write those keys.

### Design & Code Changes

- New `veomni/models/checkpoint_manager.py` — trainer-owned; caches `ParallelState` at construction (constraint 7: save/load must not follow ambient).
- Slim `checkpoint_callback.py` (`CheckpointCallback` owns DCP + HF cadence and the one-shot sidecar export); new `global_state_callback.py`.
- `BaseTrainer.checkpointer` → `BaseTrainer.checkpoint`; fan-out methods `load` / `save_dcp` / `save_hf_or_lora` / `save_model_assets`.
- Callback order: weights (`CheckpointCallback`) then cursor (`GlobalStateCallback`), so a crash between the two leaves weights whose trainer state is merely absent.
- Docs: `docs/usage/trainer.md`, `docs/key_features/lora.md`, `.agents/knowledge/{architecture,constraints}.md`.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Tests: extended an existing CI-enumerated test, **or** added a new test and wired it into `gpu_unit_tests.yml` (and `npu_unit_tests.yml` where applicable), **or** explained in **Test** why none is needed. Note that CI enumerates test files individually — a new file outside `tests/data/` and `tests/ops/` does not run until it is listed. See [.agents/knowledge/testing.md](.agents/knowledge/testing.md).

## Summary by CodeRabbit

- **New Features**
  - Unified checkpoint handling supports resumable model checkpoints and Hugging Face/LoRA exports.
  - Training position, dataloader progress, random state, and related job state are saved and restored separately.
  - Configuration and tokenizer assets are exported automatically.
  - LoRA exports use `adapter_model.safetensors` while remaining compatible with legacy PEFT formats.
  - Checkpoints from the previous format remain restorable.

- **Documentation**
  - Updated trainer, LoRA, architecture, and testing documentation to reflect the revised checkpoint workflow.

### CODE REVIEW

File: veomni/models/checkpoint_manager.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        dataloader is, the rng — has its own writer.
+        """
+        extra_state = self._extra_state(state)
+
+        helper.empty_cache()
+        self.checkpointer.save(
+            self.config.save_path,
+            {"model": self.trainer.model, "optimizer": self.trainer.optimizer, "extra_state": extra_state},
+            global_steps=state.global_step,
+            save_async=self.config.save_async,
```
[the remark this answers is not in the document]
Comment (Coach257) [reply veomni/models/checkpoint_manager.py:191]: Fixed here: 
```
def save_global_state(self, state: TrainerState) -> None:
        # Drain a pending async DCP save first. CheckpointCallback returns while
        # that write is still in flight; a cursor file that lands before the
        # shards would resume a step whose weights never made it to disk.
        checkpoint = getattr(self.trainer, "checkpoint", None)
        if checkpoint is not None:
            checkpoint.wait_for_pending_save()
```

File: .github/workflows/gpu_unit_tests.yml [hunk 2]
```diff
@@ -161,8 +161,7 @@ jobs:
           uv run --frozen pytest -s -x tests/utils/test_moe_ep_sharded_load_matrix.py
       - name: Run checkpoint callback unit tests
         run: |
-          uv run --frozen pytest -s -x tests/checkpoints/test_checkpoint_callback.py
-          uv run --frozen pytest -s -x tests/checkpoints/test_dcp_checkpointer.py
+          uv run --frozen pytest -s -x tests/checkpoints/test_checkpoint_callback.py tests/checkpoints/test_dcp_checkpointer.py
```
Comment (FoolPlayer) [inline .github/workflows/gpu_unit_tests.yml]: Can merge these in to one line `uv run --frozen pytest -s -x tests/checkpoints/`

**Review by FoolPlayer — APPROVED** [review]
